File: Data Structures & Algorithms/implement-prefix-tree/submission-9.js
class Node {
    constructor() {
        this.children = new Map() 
        this.isEnd = false
    }
} 
 
class PrefixTree {
    constructor() { 
        this.root = new Node()
    }

    /**
     * @param {string} word
     * @return {void}
     */
    insert(word) { 
        let node = this.root 

        for (const w of word) {
            if (node.children.has(w)) {
                node = node.children.get(w)
            } else {
                node.children.set(w, new Node()) 
                node = node.children.get(w)
            }
        } 

        node.isEnd = true
    }

    /**
     * @param {string} word
     * @return {boolean}
     */
    search(word) { 
        let node = this.root 

        for (const w of word) {
            if (!node.children.has(w)) {
                return false
            } else {
                node = node.children.get(w)
            }
        } 

        return node.isEnd
    }

    /**
     * @param {string} prefix
     * @return {boolean}
     */
    startsWith(prefix) { 
         let node = this.root 

        for (const w of prefix) {
            if (!node.children.has(w)) {
                return false
            } else {
                node = node.children.get(w)
            }
        }  

        return true
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-9.py
class Node: 
    def __init__(self): 
        self.desc = {} 
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        node = self.root 

        for w in word: 
            if w not in node.desc:
                node.desc[w] = Node() 

            node = node.desc[w] 

        node.isEnd = True

    def search(self, word: str) -> bool: 
        
        def dfs(i, curNode): 
            if i == len(word): 
                return curNode.isEnd 

            c = word[i] 

            if c == '.': 
                for desc in curNode.desc.values(): 
                    if dfs(i + 1, desc): 
                        return True 

                return False 
            else: 
                if c not in curNode.desc: 
                    return False 

                return dfs(i + 1, curNode.desc[c]) 

        return dfs(0, self.root)
        
